src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java

```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                "APP2.HT: Invalid number of decomposition levels");
+            isValid = false;
+        }
+
+        /* number of layers */
+
+        if (p.cod.numLayers != 1) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                String.format("APP2.HT: Number of layers (%d) is not 1", p.cod.numLayers));
+            isValid = false;
+        }
+
+        /* code-block sizes */
+
+        if (p.cod.ycb < 5 || p.cod.ycb > 6) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                String.format("APP2.HT: Invalid vertical code-block size (ycb = %d)", p.cod.ycb));
+            isValid = false;
+        }
+
+        if (p.cod.xcb < 5 || p.cod.xcb > 7) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                String.format("APP2.HT: Invalid horizontal code-block size (xcb = %d)", p.cod.xcb));
+            isValid = false;
+        }
+
+
+        /* transformation */
+
+        boolean isReversibleFilter = (p.cod.transformation == 1);
+
+        if (isHTREV && !isReversibleFilter) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                "APP2.HT: 9-7 irreversible filter is used but HTREV is signaled in CAP");
+            isValid = false;
+        }
+
+        /* precinct size */
+
+        if (p.cod.precinctSizes.length == 0 || p.cod.precinctSizes[0] != 0x77) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                "APP2.HT: Invalid precinct sizes");
+            isValid = false;
+        }
+
+        for (int i = 1; i < p.cod.precinctSizes.length; i++)
+            if (p.cod.precinctSizes[i] != 0x88) {
+                logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Invalid precinct sizes");
+                isValid = false;
+                break;
+            }
+
+        /* magbp */
+
+
+        int maxB = p.csiz[0].ssiz + 2;
+        if (isReversibleFilter) {
+            maxB += 2 + p.cod.multiComponentTransform;
+            if (p.cod.numDecompLevels > 5)
+                maxB += 1;
+        } else if (p.cod.multiComponentTransform == 1 && p.csiz[0].ssiz > 9) {
+            maxB += 1;
+        }
+
+        int codestreamB = (p.cap.ccap[0] & 0b11111) + 8;
```

Comment:
It is probably a good idea to add a comment here, that the relationship of "MAGBp" to "Parameter B" is not _always_ `+8`.
At least not in the HT-J2K spec. According to ITU-T T.814 , the table in 8.7.3 shows that the `+8` rule only works up until MAGB19. After that, the values are less consistent. Referencing the table is required.

However, this is not _currently_ an issue that Photon needs to be concerned with.  In ST 2067-21, the largest Parameter B value used, is `21`.  So for any IMF use-cases, the `MAGBp + 8` calculation will work still. But this assumption may eventually, if a future IMF update adds support for larger Parameter B values.

The code doesn't necessarily need to change. But it is probably best to document that the `codestreamB` calculation is actually based on a table in the HT-J2K spec.

Reply:
Will do.